nsqd/gossip.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	if !found {
+		n.logf("ERROR: received gossipEvent for unknown node - %s", userEv.Name)
+		return
+	}
+
+	producer := memberToProducer(member)
+	operation := userEv.Name[len(userEv.Name)-1]
+	switch operation {
+	case '+', '=':
+		n.gossipHandleCreateEvent(operation, producer, gev)
+	case '-':
+		n.gossipHandleDeleteEvent(operation, producer, gev)
+	}
+}
+
+func (n *NSQD) gossipHandleCreateEvent(operation byte,
+	producer *registrationdb.Producer, gev gossipEvent) {
+	var registrations []registrationdb.Registration
+
+	if gev.Channel != "" {
+		registrations = append(registrations, registrationdb.Registration{
+			Category: "channel",
+			Key:      gev.Topic,
+			SubKey:   gev.Channel,
+		})
+	}
+
+	registrations = append(registrations, registrationdb.Registration{
+		Category: "topic",
+		Key:      gev.Topic,
+	})
+
+	for _, r := range registrations {
+		if n.rdb.AddProducer(r, producer) {
+			n.logf("DB: member(%s) REGISTER %s", gev.Name, r)
+		}
+		if operation == '=' && n.rdb.TouchProducer(r, producer.ID) {
+			n.logf("DB: member(%s) TOUCH %s", gev.Name, r)
+		}
+	}
+}
+
+func (n *NSQD) gossipHandleDeleteEvent(operation byte,
+	producer *registrationdb.Producer, gev gossipEvent) {
+	if gev.Channel != "" {
+		r := registrationdb.Registration{
+			Category: "channel",
+			Key:      gev.Topic,
+			SubKey:   gev.Channel,
+		}
+
+		removed, left := n.rdb.RemoveProducer(r, producer.ID)
+		if removed {
+			n.logf("DB: member(%s) UNREGISTER %s", gev.Name, r)
+		}
+
+		// for ephemeral channels, if it has no producers, remove the registration
+		if left == 0 && strings.HasSuffix(gev.Channel, "#ephemeral") {
+			n.rdb.RemoveRegistration(r)
+		}
+
+		return
+	}
+
+	// this is a topic unregistration (no channel was specified)
+
+	registrations := n.rdb.FindRegistrations("channel", gev.Topic, "*")
+	for _, r := range registrations {
+		// remove all of the channel registrations...
+		if removed, _ := n.rdb.RemoveProducer(r, producer.ID); removed {
+			// normally this shouldn't happen which is why we print a warning message
+			// if anything is actually removed
+			n.logf("WARNING: client(%s) unexpected UNREGISTER %s", gev.Name, r)
+		}
+	}
+
+	r := registrationdb.Registration{
+		Category: "topic",
+		Key:      gev.Topic,
+		SubKey:   "",
+	}
+	if removed, _ := n.rdb.RemoveProducer(r, producer.ID); removed {
+		n.logf("DB: client(%s) UNREGISTER %s", gev.Name, r)
+	}
+}
+
+func (n *NSQD) serfEventLoop() {
+	for {
+		select {
+		case ev := <-n.serfEventChan:
+			switch ev.EventType() {
+			case serf.EventMemberJoin:
+				n.serfMemberJoin(ev)
+			case serf.EventMemberLeave:
+				// nothing (should never happen)
+			case serf.EventMemberFailed:
+				n.serfMemberFailed(ev)
+			case serf.EventMemberReap:
+				// nothing
+			case serf.EventUser:
+				n.serfUserEvent(ev)
+			case serf.EventQuery:
+				// nothing
+			case serf.EventMemberUpdate:
+				// nothing
+			default:
+				n.logf("WARNING: un-handled Serf event: %#v", ev)
+			}
+			gossipNotify()
```

Comment:
the "gossipNotify" thing? Yeah, that's just an unexposed mechanism to support the convergence checking in the tests. I'd be happy to improve it if you have suggestions